proposals/XXXX-extend-property-wrappers-to-functions-and-closures.md

```diff
@@ -0,0 +1,391 @@
+3. A synthesized computed property representing  `projectedValue` will be created and named per the original parameter name prefixed with a dollar sign (`$`). The accessors will mirror the `projectedValue`'s ones - except for a mutating setter which requires that the parameter be marked `inout`.
+
+#### Transformation Examples: 
+
+1. Reference Semantics Wrapper
+```swift
+@propertyWrapper
+struct Reference<Value> {
+    init(getter: () -> Value, setter: (Value) -> Void) 
+    
+    var wrappedValue: Value {
+        get 
+        nonmutating set
+    }
+    
+    var projectedValue: Self {
+        self
+    }
+}
+
+typealias A = (Reference<Int>) -> Void
+
+let a: A = { @Reference foo in
+    ...
+}
+```
+
+    Becomes:
+
+```swift
+let a: A = { _foo in
+    var foo: Int {
+        get { 
+            _foo.wrappedValue 
+        }
+        set { 
+            _foo.wrappedValue = newValue
+        }
+    }
+
+    var $foo: Int {
+        get { 
+            _foo.projectedValue 
+        }
+    }
+    
+    ...
+}
+```
+
+2. Value Semantics Wrapper
+
+```swift
+@propertyWrapper
+struct Wrapper<Value> {
+    init(wrappedValue: Value) 
+    
+    var wrappedValue: Value {
+        get 
+        set
+    }
+}
+
+typealias B = (Wrapper<Int>) -> Void
+
+let b: B = { @Wrapper foo in
+    ...
+}
+```
+
+    Becomes:
+
+```swift
+let b: B = { _foo in
+    var foo: Int {
+        _foo.wrappedValue 
+    }
+    // No setter is allowed,
+    // since foo isn't inout.
+
+    ...
+}
+```
+
+3. Value Semantics Wrapper with Special Initializer
+
+```swift
+@propertyWrapper
+class WrapperObject<Value> {
+    init(wrappedValue: Value) 
+    
+    var wrappedValue: Value {
+        get 
+        set 
+        // Not actually mutating
+        // because `WrapperObject` is
+        // a class.
+    }
+}
+
+typealias C = (WrapperObject<Int>) -> Void
+
+let c: C = { @WrapperObject foo in
+    ...
+}
+```
+
+    Becomes:
+
+```swift
+let c: C = { _foo in
+    var foo: Int {
+        get { 
+            _foo.wrappedValue 
+        }
+        set { 
+            _foo.wrappedValue = newValue
+        }
+        // Since `WrapperObject` has reference
+        // semantics we can include a setter.
+    }
+
+    ...
+}
+```
+
+
+## Source compatibility
+
+This is an additive change with _no_ impact on **source compatibility**.
+
+
+## Effect on ABI stability
+
+This is an additive change with _no_ impact on **ABI stability**.
+
+
+## Effect on API resilience
+
+This is an additive change with _no_ impact on **API resilience**.
+
+
+## Alternatives Considered
+
+**TBD**
+
+
+## Future Directions
+
+### Property Wrappers Applied to Local Declarations
+
+Should this be included? Property wrappers seem to require more structure than is provided in local variable declarations. My concern is that local property wrappers could be misused taking property wrappers' simplicity away.
```

Comment:
Local property wrappers were proposed and accepted (but not implemented) as part of the original property wrapper proposal, so they don't need to be included here. Though, the implementation for property wrapper parameters makes implementing local property wrappers pretty straightforward 🙂

Reply:
Oops, I need to freshen up on the original proposal. Though, I still feel that they're not that good of an idea.